File: interpolatepy/cubic_spline.py
import matplotlib.pyplot as plt
import numpy as np

from interpolatepy.tridiagonal_inv import solve_tridiagonal


class CubicSpline:
    """
    Cubic spline trajectory planning implementation.

    This class implements the cubic spline algorithm described in the document.
    It generates a smooth trajectory passing through specified waypoints with
    continuous velocity and acceleration profiles.
    """

    def __init__(
        self,
        t_points: list[float] | np.ndarray,
        q_points: list[float] | np.ndarray,
        v0: float = 0.0,
        vn: float = 0.0,
        debug: bool = False,
    ) -> None:
        """
        Initialize a cubic spline trajectory.

        Args:
            t_points: List or array of time points (t0, t1, ..., tn)
            q_points: List or array of position points (q0, q1, ..., qn)
            v0: Initial velocity at t0 (default: 0.0)
            vn: Final velocity at tn (default: 0.0)
            debug: Whether to print debug information (default: False)
        """
        # Ensure inputs are numpy arrays
        self.t_points = np.array(t_points, dtype=float)
        self.q_points = np.array(q_points, dtype=float)
        self.v0 = float(v0)
        self.vn = float(vn)
        self.debug = debug

        # Check input validity
        if len(self.t_points) != len(self.q_points):
            raise ValueError("Time points and position points must have the same length")

        if not np.all(np.diff(self.t_points) > 0):
            raise ValueError("Time points must be strictly increasing")

        # Compute time intervals
        self.n = len(self.t_points) - 1
        self.t_intervals = np.diff(self.t_points)

        # Compute intermediate velocities and coefficients
        self.velocities = self._compute_velocities()
        self.coefficients = self._compute_coefficients()

    def _compute_velocities(self) -> np.ndarray:
        """
        Compute the velocities at intermediate points by solving
        the tridiagonal system described in the document.

        This method implements the mathematical formulation from pages 4-5 of the document,
        corresponding to equation (17). It sets up and solves the tridiagonal system A*v = c
        to find the intermediate velocities, which ensures C2 continuity of the spline.

        Returns:
            Array of velocities [v0, v1, ..., vn]
        """
        n = self.n
        t_intervals = self.t_intervals
        q = self.q_points

        # Create the tridiagonal matrix A as shown in the document:
        # Matrix A has the following structure:
        # [2(T0+T1)    T0         0         ...        0       ]
        # [T2        2(T1+T2)     T1        0         ...      ]
        # [0          T3        2(T2+T3)    T2        ...      ]
        # [...        ...        ...        ...       ...      ]
        # [0          ...        0         Tn-2    2(Tn-3+Tn-2) Tn-3]
        # [0          ...        0          0       Tn-1      2(Tn-2+Tn-1)]

        if n == 1:
            # Special case: only one segment
            # We know v0 and vn, so no need to solve system
            return np.array([self.v0, self.vn])

        # Create the right-hand side vector c from equation (17) in the document
        # c_i = 3/(T_i*T_{i+1}) * [T_i^2*(q_{i+2}-q_{i+1}) + T_{i+1}^2*(q_{i+1}-q_i)]
        rhs = np.zeros(n - 1)

        for i in range(n - 1):
            rhs[i] = (
                3
                / (t_intervals[i] * t_intervals[i + 1])
                * (
                    t_intervals[i] ** 2 * (q[i + 2] - q[i + 1])
                    + t_intervals[i + 1] ** 2 * (q[i + 1] - q[i])
                )
            )

        # Adjust the right-hand side to account for known boundary velocities v0 and vn
        # These adjustments come from moving the terms with known velocities to the RHS
        if n > 1:
            # First equation: subtract T_1*v_0 from RHS
            rhs[0] -= t_intervals[1] * self.v0

            # Last equation: subtract T_{n-2}*v_n from RHS
            rhs[-1] -= t_intervals[-2] * self.vn

        # Solve the system for the intermediate velocities v1, ..., v(n-1)
        # Case n=2 means we have exactly one intermediate velocity to solve (1x1 system)
        if n == 2:  # noqa: PLR2004
            # Special case: only one intermediate velocity to solve for
            # Simple division is sufficient (1x1 system)
            main_diag_value = 2 * (t_intervals[0] + t_intervals[1])
            v_intermediate = rhs / main_diag_value
        else:
            # Instead of building the full matrix, extract the diagonal elements for the
            # tridiagonal solver

            # Main diagonal: 2(T_i + T_{i+1})
            main_diag = np.zeros(n - 1)
            for i in range(n - 1):
                main_diag[i] = 2 * (t_intervals[i] + t_intervals[i + 1])

            # Lower diagonal: T_{i+1} (first element not used in solve_tridiagonal)
            lower_diag = np.zeros(n - 1)
            for i in range(1, n - 1):
                lower_diag[i] = t_intervals[i + 1]

            # Upper diagonal: T_i (last element not used in solve_tridiagonal)
            upper_diag = np.zeros(n - 1)
            for i in range(n - 2):
                upper_diag[i] = t_intervals[i]

            # Print debug information only if debug is enabled
            if self.debug:
                print("\nTridiagonal Matrix components:")
                print("Main diagonal:", main_diag)
                print("Lower diagonal:", lower_diag)
                print("Upper diagonal:", upper_diag)
                print("Right-hand side vector:", rhs)

            # Solve the tridiagonal system using the Thomas algorithm
            v_intermediate = solve_tridiagonal(lower_diag, main_diag, upper_diag, rhs)

            # Print the intermediate velocities if debug is enabled
            if self.debug:
                print("\nIntermediate velocities v_1 to v_{n-1}:")
                print(v_intermediate)

        # Construct the full velocity array by including the known boundary velocities
        velocities = np.zeros(n + 1)
        velocities[0] = self.v0  # Initial velocity (given)
        velocities[1:-1] = v_intermediate  # Intermediate velocities (computed)
        velocities[-1] = self.vn  # Final velocity (given)

        # Print the complete velocity vector if debug is enabled
        if self.debug:
            print("\nComplete velocity vector v:")
            print(velocities)

        return velocities

    def _compute_coefficients(self) -> np.ndarray:
        """
        Compute the coefficients for each cubic polynomial segment.
        For each segment k, we compute:
        - ak0: constant term
        - ak1: coefficient of (t-tk)
        - ak2: coefficient of (t-tk)^2
        - ak3: coefficient of (t-tk)^3

        Returns:
            Array of shape (n, 4) containing the coefficients
        """
        n = self.n
        t_intervals = self.t_intervals
        q = self.q_points
        v = self.velocities

        coeffs = np.zeros((n, 4))

        for k in range(n):
            coeffs[k, 0] = q[k]  # ak0 = qk
            coeffs[k, 1] = v[k]  # ak1 = vk

            # Compute ak2 and ak3
            coeffs[k, 2] = (1 / t_intervals[k]) * (
                (3 * (q[k + 1] - q[k]) / t_intervals[k]) - 2 * v[k] - v[k + 1]
            )
            coeffs[k, 3] = (1 / t_intervals[k] ** 2) * (
                (2 * (q[k] - q[k + 1]) / t_intervals[k]) + v[k] + v[k + 1]
            )

            # Print detailed coefficient calculation if debug is enabled
            if self.debug:
                print(f"\nCoefficient calculation for segment {k}:")
                print(f"  ak0 = {coeffs[k, 0]}")
                print(f"  ak1 = {coeffs[k, 1]}")
                print(f"  ak2 = {coeffs[k, 2]}")
                print(f"  ak3 = {coeffs[k, 3]}")

        return coeffs

    def evaluate(self, t: float | np.ndarray) -> float | np.ndarray:
        """
        Evaluate the spline at time t.

        Args:
            t: Time point or array of time points

        Returns:
            Position(s) at the specified time(s)
        """
        t = np.atleast_1d(t)
        result = np.zeros_like(t)

        for i, ti in enumerate(t):
            # Find the segment that contains ti
            if ti <= self.t_points[0]:
                # Before the start of the trajectory
                k = 0
                tau = 0
            elif ti >= self.t_points[-1]:
                # After the end of the trajectory
                k = self.n - 1
                tau = self.t_intervals[k]
            else:
                # Within the trajectory
                # Find the largest k such that t_k <= ti
                k = np.searchsorted(self.t_points, ti, side="right") - 1
                tau = ti - self.t_points[k]

            # Evaluate the polynomial
            a = self.coefficients[k]
            result[i] = a[0] + a[1] * tau + a[2] * tau**2 + a[3] * tau**3

        return result[0] if len(result) == 1 else result

    def evaluate_velocity(self, t: float | np.ndarray) -> float | np.ndarray:
        """
        Evaluate the velocity at time t.

        Args:
            t: Time point or array of time points

        Returns:
            Velocity at the specified time(s)
        """
        t = np.atleast_1d(t)
        result = np.zeros_like(t)

        for i, ti in enumerate(t):
            # Find the segment that contains ti
            if ti <= self.t_points[0]:
                # Before the start of the trajectory
                k = 0
                tau = 0
            elif ti >= self.t_points[-1]:
                # After the end of the trajectory
                k = self.n - 1
                tau = self.t_intervals[k]
            else:
                # Within the trajectory
                k = np.searchsorted(self.t_points, ti, side="right") - 1
                tau = ti - self.t_points[k]

            # Evaluate the derivative of the polynomial
            a = self.coefficients[k]
            result[i] = a[1] + 2 * a[2] * tau + 3 * a[3] * tau**2

        return result[0] if len(result) == 1 else result

    def evaluate_acceleration(self, t: float | np.ndarray) -> float | np.ndarray:
        """
        Evaluate the acceleration at time t.

        Args:
            t: Time point or array of time points

        Returns:
            Acceleration at the specified time(s)
        """
        t = np.atleast_1d(t)
        result = np.zeros_like(t)

        for i, ti in enumerate(t):
            # Find the segment that contains ti
            if ti <= self.t_points[0]:
                # Before the start of the trajectory
                k = 0
                tau = 0
            elif ti >= self.t_points[-1]:
                # After the end of the trajectory
                k = self.n - 1
                tau = self.t_intervals[k]
            else:
                # Within the trajectory
                k = np.searchsorted(self.t_points, ti, side="right") - 1
                tau = ti - self.t_points[k]

            # Evaluate the second derivative of the polynomial
            a = self.coefficients[k]
            result[i] = 2 * a[2] + 6 * a[3] * tau

        return result[0] if len(result) == 1 else result

    def plot(self, num_points: int = 1000) -> None:
        """
        Plot the spline trajectory along with its velocity and acceleration profiles.

        Args:
            num_points: Number of points to use for plotting
        """
        t_min, t_max = self.t_points[0], self.t_points[-1]
        t = np.linspace(t_min, t_max, num_points)

        q = self.evaluate(t)
        v = self.evaluate_velocity(t)
        a = self.evaluate_acceleration(t)

        _fig, (ax1, ax2, ax3) = plt.subplots(3, 1, figsize=(10, 8), sharex=True)

        # Position plot
        ax1.plot(t, q, "b-", linewidth=2)
        ax1.plot(self.t_points, self.q_points, "ro", markersize=8)
        ax1.set_ylabel("Position")
        ax1.grid(True)
        ax1.set_title("Cubic Spline Trajectory")

        # Velocity plot
        ax2.plot(t, v, "g-", linewidth=2)
        ax2.plot(self.t_points, self.velocities, "ro", markersize=6)
        ax2.set_ylabel("Velocity")
        ax2.grid(True)

        # Acceleration plot
        ax3.plot(t, a, "r-", linewidth=2)
        ax3.set_ylabel("Acceleration")
        ax3.set_xlabel("Time")
        ax3.grid(True)

        plt.tight_layout()
        plt.show()

File: cpp/src/cubic_spline.cpp
#include <interpolatecpp/spline/cubic_spline.hpp>
#include <interpolatecpp/tridiagonal.hpp>

#include <algorithm>
#include <cmath>
#include <iostream>
#include <stdexcept>

namespace interpolatecpp::spline {

CubicSpline::CubicSpline(std::span<const double> t_points, std::span<const double> q_points,
                          double v0, double vn, bool debug)
    : v0_(v0), vn_(vn), debug_(debug) {
    if (t_points.size() != q_points.size()) {
        throw std::invalid_argument("Time points and position points must have the same length");
    }

    const auto n_points = static_cast<int>(t_points.size());
    if (n_points < 2) {
        throw std::invalid_argument("At least two points are required");
    }

    t_points_ = Eigen::Map<const Eigen::VectorXd>(t_points.data(), n_points);
    q_points_ = Eigen::Map<const Eigen::VectorXd>(q_points.data(), n_points);

    // Check strictly increasing
    for (int i = 1; i < n_points; ++i) {
        if (t_points_[i] <= t_points_[i - 1]) {
            throw std::invalid_argument("Time points must be strictly increasing");
        }
    }

    n_ = n_points - 1;
    t_intervals_ = t_points_.tail(n_).array() - t_points_.head(n_).array();

    compute_velocities();
    compute_coefficients();
}

void CubicSpline::compute_velocities() {
    if (n_ == 1) {
        velocities_.resize(2);
        velocities_(0) = v0_;
        velocities_(1) = vn_;
        return;
    }

    // Build RHS: c_i = 3/(T_i*T_{i+1}) * [T_i^2*(q_{i+2}-q_{i+1}) + T_{i+1}^2*(q_{i+1}-q_i)]
    const int m = n_ - 1;  // number of intermediate velocities
    Eigen::VectorXd rhs(m);

    for (int i = 0; i < m; ++i) {
        rhs(i) = 3.0 / (t_intervals_(i) * t_intervals_(i + 1)) *
                 (t_intervals_(i) * t_intervals_(i) * (q_points_(i + 2) - q_points_(i + 1)) +
                  t_intervals_(i + 1) * t_intervals_(i + 1) * (q_points_(i + 1) - q_points_(i)));
    }

    // Adjust for boundary velocities
    rhs(0) -= t_intervals_(1) * v0_;
    rhs(m - 1) -= t_intervals_(n_ - 2) * vn_;

    Eigen::VectorXd v_intermediate;

    if (n_ == 2) {
        // 1x1 system: simple division
        double main_diag_value = 2.0 * (t_intervals_(0) + t_intervals_(1));
        v_intermediate = rhs / main_diag_value;
    } else {
        // Build tridiagonal diagonals
        Eigen::VectorXd main_diag(m);
        Eigen::VectorXd lower_diag = Eigen::VectorXd::Zero(m);
        Eigen::VectorXd upper_diag = Eigen::VectorXd::Zero(m);

        for (int i = 0; i < m; ++i) {
            main_diag(i) = 2.0 * (t_intervals_(i) + t_intervals_(i + 1));
        }
        for (int i = 1; i < m; ++i) {
            lower_diag(i) = t_intervals_(i + 1);
        }
        for (int i = 0; i < m - 1; ++i) {
            upper_diag(i) = t_intervals_(i);
        }

        if (debug_) {
            std::cout << "\nTridiagonal Matrix components:\n";
            std::cout << "Main diagonal: " << main_diag.transpose() << "\n";
            std::cout << "Lower diagonal: " << lower_diag.transpose() << "\n";
            std::cout << "Upper diagonal: " << upper_diag.transpose() << "\n";
            std::cout << "Right-hand side vector: " << rhs.transpose() << "\n";
        }

        v_intermediate = solve_tridiagonal(lower_diag, main_diag, upper_diag, rhs);

        if (debug_) {
            std::cout << "\nIntermediate velocities: " << v_intermediate.transpose() << "\n";
        }
    }

    // Assemble full velocity vector
    velocities_.resize(n_ + 1);
    velocities_(0) = v0_;
    velocities_.segment(1, m) = v_intermediate;
    velocities_(n_) = vn_;

    if (debug_) {
        std::cout << "\nComplete velocity vector: " << velocities_.transpose() << "\n";
    }
}

void CubicSpline::compute_coefficients() {
    coefficients_.resize(n_, 4);

    for (int k = 0; k < n_; ++k) {
        double T = t_intervals_(k);
        double q_k = q_points_(k);
        double q_k1 = q_points_(k + 1);
        double v_k = velocities_(k);
        double v_k1 = velocities_(k + 1);

        coefficients_(k, 0) = q_k;                                                       // a0
        coefficients_(k, 1) = v_k;                                                       // a1
        coefficients_(k, 2) = (1.0 / T) * (3.0 * (q_k1 - q_k) / T - 2.0 * v_k - v_k1); // a2
        coefficients_(k, 3) =
            (1.0 / (T * T)) * (2.0 * (q_k - q_k1) / T + v_k + v_k1);                    // a3

        if (debug_) {
            std::cout << "\nCoefficient calculation for segment " << k << ":\n";
            std::cout << "  a0 = " << coefficients_(k, 0) << "\n";
            std::cout << "  a1 = " << coefficients_(k, 1) << "\n";
            std::cout << "  a2 = " << coefficients_(k, 2) << "\n";
            std::cout << "  a3 = " << coefficients_(k, 3) << "\n";
        }
    }
}

CubicSpline::SegmentInfo CubicSpline::find_segment(double t) const {
    if (t <= t_points_(0)) {
        return {0, 0.0};
    }
    if (t >= t_points_(n_)) {
        return {n_ - 1, t_intervals_(n_ - 1)};
    }
    // Binary search: find largest k such that t_points_[k] <= t
    auto begin = t_points_.data();
    auto end = begin + n_ + 1;
    auto it = std::upper_bound(begin, end, t);
    int k = static_cast<int>(it - begin) - 1;
    return {k, t - t_points_(k)};
}

double CubicSpline::evaluate(double t) const {
    auto [k, tau] = find_segment(t);
    double a0 = coefficients_(k, 0);
    double a1 = coefficients_(k, 1);
    double a2 = coefficients_(k, 2);
    double a3 = coefficients_(k, 3);
    return a0 + a1 * tau + a2 * tau * tau + a3 * tau * tau * tau;
}

Eigen::VectorXd CubicSpline::evaluate(const Eigen::VectorXd& t) const {
    Eigen::VectorXd result(t.size());
    for (Eigen::Index i = 0; i < t.size(); ++i) {
        result(i) = evaluate(t(i));
    }
    return result;
}

double CubicSpline::evaluate_velocity(double t) const {
    auto [k, tau] = find_segment(t);
    double a1 = coefficients_(k, 1);
    double a2 = coefficients_(k, 2);
    double a3 = coefficients_(k, 3);
    return a1 + 2.0 * a2 * tau + 3.0 * a3 * tau * tau;
}

Eigen::VectorXd CubicSpline::evaluate_velocity(const Eigen::VectorXd& t) const {
    Eigen::VectorXd result(t.size());
    for (Eigen::Index i = 0; i < t.size(); ++i) {
        result(i) = evaluate_velocity(t(i));
    }
    return result;
}

double CubicSpline::evaluate_acceleration(double t) const {
    auto [k, tau] = find_segment(t);
    double a2 = coefficients_(k, 2);
    double a3 = coefficients_(k, 3);
    return 2.0 * a2 + 6.0 * a3 * tau;
}

Eigen::VectorXd CubicSpline::evaluate_acceleration(const Eigen::VectorXd& t) const {
    Eigen::VectorXd result(t.size());
    for (Eigen::Index i = 0; i < t.size(); ++i) {
        result(i) = evaluate_acceleration(t(i));
    }
    return result;
}

}  // namespace interpolatecpp::spline
